cms/server/admin/handlers/contestranking.py

```diff
@@ -169,3 +225,237 @@ def _status_indicator(status: TaskStatus) -> str:
         if not status.has_opened:
             return "!" + star
         return star
+
+
+class ScoreHistoryHandler(BaseHandler):
+    """Returns the score history for a contest as JSON.
+
+    This endpoint provides score history data in RWS format:
+    [[user_id, task_id, time, score], ...]
+
+    This matches the format expected by RWS's HistoryStore.js for
+    computing score and rank histories.
+
+    By default, excludes hidden participations to match ranking page behavior.
+    Use ?include_hidden=1 to include hidden participations.
+
+    Before returning history data, this handler checks for any cache entries
+    with history_valid=False and rebuilds their history to ensure correctness.
+
+    """
+    @require_permission(BaseHandler.AUTHENTICATED)
+    def get(self, contest_id):
+        contest = self.safe_get_item(Contest, contest_id)
+
+        include_hidden = self.get_argument("include_hidden", "0") == "1"
+
+        invalid_score_entries = (
+            self.sql_session.query(ParticipationTaskScore)
+            .join(Participation)
+            .filter(Participation.contest_id == contest_id)
+            .filter(ParticipationTaskScore.score_valid.is_(False))
+            .options(joinedload(ParticipationTaskScore.participation))
+            .options(joinedload(ParticipationTaskScore.task))
+            .all()
+        )
+
+        for entry in invalid_score_entries:
+            rebuild_score_cache(
+                self.sql_session, entry.participation, entry.task
+            )
+
+        invalid_history_entries = (
+            self.sql_session.query(ParticipationTaskScore)
+            .join(Participation)
+            .filter(Participation.contest_id == contest_id)
+            .filter(ParticipationTaskScore.score_valid.is_(True))
+            .filter(ParticipationTaskScore.history_valid.is_(False))
+            .options(joinedload(ParticipationTaskScore.participation))
+            .options(joinedload(ParticipationTaskScore.task))
+            .all()
+        )
+
+        for entry in invalid_history_entries:
+            rebuild_score_history(
+                self.sql_session, entry.participation, entry.task
+            )
+
+        if invalid_score_entries or invalid_history_entries:
+            self.sql_session.commit()
+
+        query = (
+            self.sql_session.query(ScoreHistory)
+            .join(Participation)
+            .filter(Participation.contest_id == contest_id)
+            .options(joinedload(ScoreHistory.participation).joinedload(Participation.user))
+        )
+
+        if not include_hidden:
+            query = query.filter(Participation.hidden.is_(False))
+
+        history = query.order_by(ScoreHistory.timestamp).all()
```

Comment:
Could this be factored to a function in scorecache.py like get_cached_score_entry? This seems fragile